packages/api/lambdas/publish-reports.js

```diff
@@ -0,0 +1,267 @@
+ * @param {string} [granuleSnsTopicArn]
+ *   SNS topic ARN for reporting granules. Defaults to `process.env.granule_sns_topic_arn`.
+ * @returns {Promise}
+ */
+async function publishGranuleSnsMessage(
+  granuleRecord,
+  granuleSnsTopicArn = process.env.granule_sns_topic_arn
+) {
+  return publishSnsMessage(granuleSnsTopicArn, granuleRecord);
+}
+
+/**
+ * Publish SNS message for PDR reporting.
+ *
+ * @param {Object} pdrRecord - A PDR record.
+ * @param {string} [pdrSnsTopicArn]
+ *   SNS topic ARN for reporting PDRs. Defaults to `process.env.pdr_sns_topic_arn`.
+ * @returns {Promise}
+ */
+async function publishPdrSnsMessage(
+  pdrRecord,
+  pdrSnsTopicArn = process.env.pdr_sns_topic_arn
+) {
+  return publishSnsMessage(pdrSnsTopicArn, pdrRecord);
+}
+
+/**
+ * Publish execution record to SNS topic.
+ *
+ * @param {Object} eventMessage - Workflow execution message
+ * @returns {Promise}
+ */
+async function handleExecutionMessage(eventMessage) {
+  try {
+    const executionRecord = await Execution.generateRecord(eventMessage);
+    return publishExecutionSnsMessage(executionRecord);
+  } catch (err) {
+    log.error('Error handling execution message', err);
+    log.info('Execution message', eventMessage);
+    return Promise.resolve();
+  }
+}
+
+/**
+ * Build a granule record and publish it to SNS for granule reporting.
+ *
+ * @param {Object} granule - A granule object
+ * @param {Object} eventMessage - A workflow execution message
+ * @param {string} executionUrl - A Step Function execution URL
+ * @param {Object} [executionDescription={}] - Defaults to empty object
+ * @param {Date} executionDescription.startDate - Start date of the workflow execution
+ * @param {Date} executionDescription.stopDate - Stop date of the workflow execution
+ * @returns {Promise}
+ */
+async function buildAndPublishGranule(
+  granule,
+  eventMessage,
+  executionUrl,
+  executionDescription = {}
+) {
+  try {
+    const granuleRecord = await Granule.generateGranuleRecord(
+      granule,
+      eventMessage,
+      executionUrl,
+      executionDescription
+    );
+    return publishGranuleSnsMessage(granuleRecord);
+  } catch (err) {
+    log.error('Error handling granule from message', err);
+    log.info('Granule from message', granule);
+    log.info('Execution message', eventMessage);
+    return Promise.resolve();
+  }
+}
+
+/**
+ * Publish individual granule messages to SNS topic.
+ *
+ * @param {Object} eventMessage - Workflow execution message
+ * @returns {Promise}
+ */
+async function handleGranuleMessages(eventMessage) {
+  const granules = getMessageGranules(eventMessage);
+  if (!granules) {
+    log.info('No granules to process on the message');
+    return Promise.resolve();
+  }
+
+  const executionArn = getMessageExecutionArn(eventMessage);
+  const executionUrl = getExecutionUrl(executionArn);
+
+  let executionDescription;
+  try {
+    executionDescription = await StepFunctions.describeExecution({ executionArn });
```

Comment:
The reason for the extra try/catch here is that I decided that a failure in describing the step function execution (which could happen for any number of reasons) shouldn't prevent the granule from being published to the SNS topic, which it would if I included it in the below try/catch block.